### ISSUE TITLE
Unable to run WinRM tasks since 1.9.5, KeyError on self.delegate

### ISSUE THREAD

**Issue by shaunbrady** [ansible/ansible#15153]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

```
<!--- Paste verbatim output from “ansible --version” between quotes -->
ansible 1.9.5
  configured module search path = None
```
##### CONFIGURATION

```
[defaults]
nocows=1

[ssh_connection]
pipelining = True
```
##### OS / ENVIRONMENT

Tested on Ubuntu 14.04.2 LTS, managing Windows 2012
##### SUMMARY

I am unable to run WinRM tasks since upgrade Ansible to 1.9.5 (PyPI)
##### STEPS TO REPRODUCE

```
  1 --                                                                                                                                                                                                 
  2 - name: Create the Windows 2012 R2 EC2 instance
  3   hosts: localhost
  4   connection: local
  5   vars_files:
  6     - vars/common-gov.encrypted.yaml
  7   roles:
  8     - role: ec2-instance-create
  9       #### sets variable ec2_instance
 10   tasks:
 11     - name: Get the Administrator password
 12       ec2_win_password:
 13         instance_id: "{{ ec2_instance.id }}"
 14         region: "us-gov-west-1"
 15         key_file: "keys/ansible.encrypted"
 16         wait: yes
 17         wait_timeout: 45
 18       register: win_pass
 19     - name: Add host to platforms group
 20       add_host:
 21         hostname="{{ ec2_instance.public_ip }}"
 22         ansible_ssh_user="Administrator"
 23         ansible_ssh_pass="{{ win_pass.win_password }}"
 24         ansible_ssh_port=5986
 25         ansible_connection=winrm
 26         ansible_winrm_server_cert_validation=ignore
 27         groups=platforms,windows
 28 
 29 - name: Test
 30   hosts: platforms
 31   gather_facts: False
 32   tasks:
 33     - name: Put PuTTY in place
 34       win_get_url:
 35         url: http://the.earth.li/~sgtatham/putty/latest/x86/putty.exe
 36         dest: C:\Users\Public\Desktop\putty.exe
```

##### EXPECTED RESULTS

I expected the playbook to complete.
##### ACTUAL RESULTS

```
PLAY [Test] ******************************************************************* 

TASK: [Put PuTTY in place] **************************************************** 
fatal: [96.127.111.61] => Traceback (most recent call last):
  File "/home/sbrady/.virtualenvs/nimbis-platform/local/lib/python2.7/site-packages/ansible/runner/__init__.py", line 589, in _executor
    exec_rc = self._executor_internal(host, new_stdin)
  File "/home/sbrady/.virtualenvs/nimbis-platform/local/lib/python2.7/site-packages/ansible/runner/__init__.py", line 793, in _executor_internal
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)
  File "/home/sbrady/.virtualenvs/nimbis-platform/local/lib/python2.7/site-packages/ansible/runner/__init__.py", line 972, in _executor_internal_inner
    conn = self.connector.connect(actual_host, actual_port, actual_user, actual_pass, actual_transport, actual_private_key_file, delegate_host)
  File "/home/sbrady/.virtualenvs/nimbis-platform/local/lib/python2.7/site-packages/ansible/runner/connection.py", line 52, in connect
    self.active = conn.connect()
  File "/home/sbrady/.virtualenvs/nimbis-platform/local/lib/python2.7/site-packages/ansible/runner/connection_plugins/winrm.py", line 155, in connect
    self.protocol = self._winrm_connect()
  File "/home/sbrady/.virtualenvs/nimbis-platform/local/lib/python2.7/site-packages/ansible/runner/connection_plugins/winrm.py", line 81, in _winrm_connect
    host_vars = self.runner.inventory._hosts_cache[self.delegate].get_variables()
KeyError: None


FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/sbrady/ec2-windows.yaml.retry

```

**Comment by nitzmahone** [comment]
I'm not able to reproduce this on 1.9.5 with a normal installation from PyPI. Running Ansible from a virtualenv is fraught with peril- I do it a lot for development, and I trip over virtualenv-specific issues constantly (due to internal use of subprocesses that don't always inherit the parent environment [and the virtualenv settings]). I suspect what you're seeing is the result of a mixed installation, where the top-level runner executes in the virtualenv and subprocesses inherit the original system Python environment. It causes lots of issues when those two envs don't match.

Can you try this from a standard non-virtualized installation and see if you get the same results?

**Comment by nitzmahone** [comment]
@shaunbrady Please let us know if you're still experiencing this issue after trying what we've requested above. Closing for now...

**Comment by shaunbrady** [comment]
Didn't have time this weekend.  We try this coming to get a successful result.  Thank you.

**Comment by jinzishuai** [comment]
We have exactly the same problem. 
We installed ansible using the `ppa:ansible/ansible-1.9` on Ubuntu-14.04. 

Is there any way to go back to **1.9.4** which worked great?

**Comment by ghost** [comment]
We are also seeing this. Out of nowhere, on my machine only - winrm is failing in the same way. I upgraded ansible via pip to 1.9.6, same error. Downgraded to 1.9.4, works like a charm. No virtualenv involved in our environment and this seemed to happen out of the blue.

**Comment by nitzmahone** [comment]
Are you doing anything funky with your inventory or virtualenvs? Which auth type are you using? Nobody's been able to share a reproduction of this, and it works fine here...

**Comment by kosch** [comment]
I've the same issue as I tried today windows integration with 1.9.6. With 1.9.4 it works fine. ubuntu 16.04
`
fatal: [selenium-win7-ci-n1] => Traceback (most recent call last):
  File "/usr/local/lib/python2.6/dist-packages/ansible/runner/__init__.py", line 589, in _executor                                                                                                                              
    exec_rc = self._executor_internal(host, new_stdin)                                                                                                                                                                          
  File "/usr/local/lib/python2.6/dist-packages/ansible/runner/__init__.py", line 793, in _executor_internal                                                                                                                     
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)                                                                                                     
  File "/usr/local/lib/python2.6/dist-packages/ansible/runner/__init__.py", line 972, in _executor_internal_inner                                                                                                               
    conn = self.connector.connect(actual_host, actual_port, actual_user, actual_pass, actual_transport, actual_private_key_file, delegate_host)                                                                                 
  File "/usr/local/lib/python2.6/dist-packages/ansible/runner/connection.py", line 52, in connect                                                                                                                               
    self.active = conn.connect()                                                                                                                                                                                                
  File "/usr/local/lib/python2.6/dist-packages/ansible/runner/connection_plugins/winrm.py", line 155, in connect                                                                                                                
    self.protocol = self._winrm_connect()                                                                                                                                                                                       
  File "/usr/local/lib/python2.6/dist-packages/ansible/runner/connection_plugins/winrm.py", line 81, in _winrm_connect                                                                                                          
    host_vars = self.runner.inventory._hosts_cache[self.delegate].get_variables()                                                                                                                                               
KeyError: None   
`

**Comment by ghost** [comment]
We're still seeing this as well. I experimented with this a bit the other day. I tried with a flat inventory with no child groups. WinRM is set to password authentication over SSL. We're not using virtualenv at all. I've tried pretty much every version of pywinrm on pypi for both ansible 1.9.5 and 1.9.6, nothing seems to be working.

I've gone so far as: installing new ubuntu 14.04.4 vm, python 2.7.6 (ubuntu 14.04 default), pip install ansible==1.9.6 pywinrm==0.1.1 and running the sample win_ping example as laid out in the documentation: http://docs.ansible.com/ansible/intro_windows.html

I still get this error. I also tried the same test with python 2.7.9 in a virtualenv, same results.

`ubuntu@ubuntu:~$ ansible windows -i inventory -m win_ping
172.26.23.8 | FAILED => Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/ansible/runner/__init__.py", line 589, in _executor
    exec_rc = self._executor_internal(host, new_stdin)
  File "/usr/local/lib/python2.7/dist-packages/ansible/runner/__init__.py", line 793, in _executor_internal
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)
  File "/usr/local/lib/python2.7/dist-packages/ansible/runner/__init__.py", line 972, in _executor_internal_inner
    conn = self.connector.connect(actual_host, actual_port, actual_user, actual_pass, actual_transport, actual_private_key_file, delegate_host)
  File "/usr/local/lib/python2.7/dist-packages/ansible/runner/connection.py", line 52, in connect
    self.active = conn.connect()
  File "/usr/local/lib/python2.7/dist-packages/ansible/runner/connection_plugins/winrm.py", line 155, in connect
    self.protocol = self._winrm_connect()
  File "/usr/local/lib/python2.7/dist-packages/ansible/runner/connection_plugins/winrm.py", line 81, in _winrm_connect
    host_vars = self.runner.inventory._hosts_cache[self.delegate].get_variables()
KeyError: None
`

**Comment by nitzmahone** [comment]
@seanpowell-motiga  @kosch Any reason you're still on such an old version of Ansible? At this point, even if we could repro the issue locally, it's extremely unlikely that we'll be patching Ansible 1.9 for anything other than egregious security issues. You're basically three major releases back (with 2.2 currently at release candidate). There's a lot of performance and module goodies on the Windows stuff that you're missing out on by staying on such an old release.

**Comment by ghost** [comment]
I've resorted patching winrm.py in ansible directly. This is not ideal, obviously, but it stops me from being blocked by this issue. May be helpful to others.

Attached is my patch file.

[ansible-winrm-patch.txt](https://github.com/ansible/ansible/files/528240/ansible-winrm-patch.txt)

**Comment by ghost** [comment]
We're in the process of upgrading to 2.2, but that is rather involved for our setup. Ansible 2.x is not as backwards compatible as it seems to be made out to be. We need at least 1.9.5 because of our requirements with newer versions of Docker.

**Comment by bcoca** [comment]
There is a migration guide from 1.9 to 2.0 but 2.2 is removing things we deprecated in 2.0, the more you wait the harder the upgrade.

**Comment by ghost** [comment]
Oh, I'm not waiting. I just can't roll 2.2 out today, but I need the new docker stuff today. It's all a matter of staffing/time to do the upgrade. I need more of both :)

I have another fix:

file: ansible/runner/connection_plugins/winrm.py
from:
Line 81: host_vars = self.runner.inventory._hosts_cache[self.delegate].get_variables()

to:
Line 81: host_vars = self.runner.inventory._hosts_cache[self.host].get_variables()

Seems to also fix it (and seemingly in a better way). self.delegate is always None as far as I can tell, not sure how that could have possibly ever worked. 

I understand you guys don't want to patch older versions, but as there was a new response today there are clearly more people being affected by this, so as long as the community can help each other I'm all for that.

**Comment by kosch** [comment]
@nitzmahone I tried to migrate to 2.0 one year ago, but cancelled this process due lots of bugs in 2.0 and many incompatbilities especially at include-behaviour. Few days ago I tried 2.1, it looked better but there are still some issues which makes the migration a more time consuming process than expected.
One major thing is also the change of task name rendering in 2.x. In 1.x I can see what happens more or lesse detailed, which is extremly helpful finding causes of errors whil playbook runs. in 2.x I see only e.g.
TASK: [ copy ] *****
No details what is copied to where... I did not find an option to enable the 1.x behaviour for the default names of tasks and writing a name for every task by copy its task action text, just to see what happens here, is an useless overhead.
